### PR TITLE
This commit adds the initial server and rendering components.

### PR DESCRIPTION
**Pull request by EthanSteinberg**
It results in a simple game where you can use the arrow keys to move your block around
the map.

Note: This suffers from severe lag problems and/or random crashes.

This change is [Reviewable](https://reviewable.io/reviews/loungecpp/tatsoryk/20)

### CODE REVIEW

**Comment by TheCatPlusPlus** [comment]
I don't see the point of merging in code that will be immediately removed anyway (all the temporary protocol bits etc). Rebase this on nab's branch and we'll pull it after that is reviewed.

Also @thecoshman is doing server loop I think so needs to be reconciled with whatever he's got.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

_[bin/provision, line 21 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE552O1ox-Bw9YC-iNx:-KE552O1ox-Bw9YC-iNy:-1113465608) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/bin/provision#L21)):_
This shouldn't be needed, and may cause problems. Make the server ignore the path instead.

---

_[client/game.js, line 79 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE55L10KQbAlkxHY9KT:-KE55L10KQbAlkxHY9KU:-1611618572) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/client/game.js#L79)):_
Time delta is necessary for interpolation, don't remove this.

---

_[server/src/main.rs, line 52 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE56j5hql65RBTh1QpU:-KE56j5iaZKHdKfPx6ZR:-1513978149) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/server/src/main.rs#L52)):_
All the protocol bits should use nab's code.

---

_[server/src/main.rs, line 86 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE576qTEBTuY6hHLl1T:-KE576qTEBTuY6hHLl1U:-1854319368) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/server/src/main.rs#L86)):_
No unwraps.

---

_[server/src/main.rs, line 241 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE59PHMMuUYEa9RKHVg:-KE59PHMMuUYEa9RKHVh:1704508498) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/server/src/main.rs#L241)):_
After 20 seconds the connections should be long dropped due to inactivity. Server loop can run at around the same FPS as the clients (i.e. ~60), the calculations are not very involved.

---

_[server/src/main.rs, line 273 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE59yVm7e8CZKalTXSZ:-KE59yVm7e8CZKalTXS_:-620190601) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/server/src/main.rs#L273)):_
No unwraps. Clients that can't be read from should just be dropped.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-:-KE56-DHx2N1HX9h7Ot_:-2064522239)_

**Comment by EthanSteinberg** [comment]
@thecoshman decided to drink yesterday instead of code and allowed me to code these parts in. I was originally planning to rebase on nab's branch, but he told me not to.  I'll do it right now.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

_[bin/provision, line 21 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE3fyiUWO7fhyPPOHTr:-KE3fyiVAo2ZVjTG1JmK:-768231350) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e8047789d38b36bdd70daf2113b7dc4b62fb0012/bin/provision#L21)):_
This is necessary to prevent nginx from sending redirects.

---

_[server/src/main.rs, line 52 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE56j5hql65RBTh1QpU:-KE6YApNs9g8nA-xWJGP:-1730282996) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/server/src/main.rs#L52)):_
Yeah these should probably be renamed. They are WebsocketEvents orGameEvents , not really messages. They are the means for transporting data from the websocket threads to the game thread.

---

_[server/src/main.rs, line 86 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE576qTEBTuY6hHLl1T:-KE6YVigHNzYuCW-JIZ9:-1615243884) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/server/src/main.rs#L86)):_
Sure. I'll swap these out.

---

_[server/src/main.rs, line 273 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE59yVm7e8CZKalTXSZ:-KE6Ye9iFFPuqKr5Hvnq:-2000384973) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/server/src/main.rs#L273)):_
Sure thing. I'll corresponding drop logic to the read loops.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-:-KE6XtO7MX1PLuziNHiF:1971215576)_

**Comment by EthanSteinberg** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

_[bin/provision, line 21 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE552O1ox-Bw9YC-iNx:-KE6YucxfJ8auJyMUC_Z:-438043289) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/bin/provision#L21)):_
? I am not sure exactly what you mean by this? Or how to do it with nginx.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20)_

**Comment by EthanSteinberg** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

_[server/src/main.rs, line 273 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE59yVm7e8CZKalTXSZ:-KE6cYRxAxuoaA-3ZY9z:2032523232) ([raw file](https://github.com/loungecpp/tatsoryk/blob/e19ca194b5c561bf615703e28c9d0b89664dea82/server/src/main.rs#L273)):_
Actually, this whole handle_connection logic is done in it's own thread, so I think it might be cleaner to just panic and have that single thread shut down.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20)_

**Comment by TheCatPlusPlus** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

_[bin/provision, line 21 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE552O1ox-Bw9YC-iNx:-KE6hgHzJbvnNmb1fdXW:-1479381369) ([raw file](https://github.com/loungecpp/tatsoryk/blob/41c44dcd004be532b54720d66c742c122333d9fb/bin/provision#L21)):_
This will match all URLs that begin with `ws`, even files etc. It could be `= /ws` to match exactly, but you can just use `ws://localhost/ws/` in the client to avoid the redirect, so whichever you prefer I guess. This endpoint will need some discovery protocol later anyway.

---

_[server/src/main.rs, line 52 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE56j5hql65RBTh1QpU:-KE6iGOfKuCXD-YzmpxF:363610133) ([raw file](https://github.com/loungecpp/tatsoryk/blob/41c44dcd004be532b54720d66c742c122333d9fb/server/src/main.rs#L52)):_
Oh, okay, derp.

---

_[server/src/main.rs, line 273 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE59yVm7e8CZKalTXSZ:-KE6imUHbnFCrKS1MMwk:176117215) ([raw file](https://github.com/loungecpp/tatsoryk/blob/41c44dcd004be532b54720d66c742c122333d9fb/server/src/main.rs#L273)):_
I think it's better to have `try!` here anyway, and if we really want a panic then add a single unwrap in the caller.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20)_

File: server/src/main.rs [hunk 1]
```diff
@@ -6,16 +6,350 @@
     unused_import_braces,
     unused_qualifications,
     unsafe_code,
-    dead_code,
-    unused_results,
```
Comment (nabijaczleweli) [inline server/src/main.rs:10]: Don't

**Comment by nabijaczleweli** [comment]
y u no docs?

File: server/src/main.rs (hunk 1 above)
Comment (thecoshman) [inline server/src/main.rs:10]: If nab's is saying that we should keep this as compilation errors, then I agree.

File: server/src/main.rs [hunk 2]
```diff
@@ -10,17 +10,345 @@
...
+*/
+#[derive(Debug)]
+enum WebSocketEvent {
+    ClientCreated {client: Client},
+    ClientClosed {client_id: u32},
+    ClientMessage {client_id: u32, message: message::Message},
+}
+
+/*
+  This represends a single websocket connected to the game.
```
Comment (thecoshman) [inline server/src/main.rs]: spelling: represents

File: server/src/main.rs [hunk 3]
```diff
@@ -10,17 +10,345 @@
...
+  The id is simply the ip address and port.
```
Comment (thecoshman) [inline server/src/main.rs]: id should probably be uuid, but we can change that later

File: server/src/main.rs [hunk 4]
```diff
@@ -10,17 +10,345 @@
...
+  'sender' is a channel which allows you to send messages to the corresponding websocket.
+  Send a None to close the websocket. (Some(data) for a normal message).
```
Comment (thecoshman) [inline server/src/main.rs]: Not sure I like the idea of this... why not a proper 'close' function?

File: server/src/main.rs [hunk 5]
```diff
@@ -10,17 +10,345 @@
...
+impl Client {
+    /* 
+      Create a new client from a given id and sender channel.
+    */
+    fn new(id: u32, sender: std::sync::mpsc::Sender<Option<String>>) -> Client {
+        let result = Client {
+            id: id,
+            sender: sender,
+        };
+        result
```
Comment (thecoshman) [inline server/src/main.rs]: I think this line is redundant...

**Comment by EthanSteinberg** [comment]
Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

_[server/src/main.rs, line 52 [r5]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE3fZ42sDHZxHZPgkHG-r5-52:-KE7JbAbJLue-bUZSAng:299737818) ([raw file](https://github.com/loungecpp/tatsoryk/blob/18cdc5e847260958952d19a2954d7b2aede607f9/server/src/main.rs#L52)):_
There is a close function. That's an implementation detail. I'll move that comment out of the public docs.

---

_[server/src/main.rs, line 68 [r5]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE3fZ42sDHZxHZPgkHG-r5-68:-KE7JmuP_iL_fztpgPO4:1350736562) ([raw file](https://github.com/loungecpp/tatsoryk/blob/18cdc5e847260958952d19a2954d7b2aede607f9/server/src/main.rs#L68)):_
Yep. Got missed in a refactor.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20)_

**Comment by EthanSteinberg** [comment]
Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

_[server/src/main.rs, line 49 [r5]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20#-KE3fZ42sDHZxHZPgkHG-r5-49:-KE7KCLILWihffmtAQIT:-499146009) ([raw file](https://github.com/loungecpp/tatsoryk/blob/18cdc5e847260958952d19a2954d7b2aede607f9/server/src/main.rs#L49)):_
Done.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20)_

File: server/src/gamestate.rs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+use std::iter::FromIterator;
+
+use events::Client;
+use events::WebSocketEvent;
+
+/// The GameState contains the whole state of the game.
+///
+/// It consists of both players, and all the clients which are currently connected.
+#[derive(Debug)]
+pub struct GameState {
```
Comment (thecoshman) [inline server/src/gamestate.rs:19]: What about bullets?

File: server/src/gamestate.rs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    /// Updates the game state in one tick.
+    pub fn process_game_update(&mut self) {
+        for (_, player) in &mut self.players {
+            player.x += player.move_x.unwrap_or(0.0);
+            player.y += player.move_y.unwrap_or(0.0);
+        }
+    }
+
+    /// Send the current state to each client.
+    pub fn send_state_updates(&self) {
```
Comment (thecoshman) [inline server/src/gamestate.rs:53]: Are we doing this _every_ game tick?

File: server/src/gamestate.rs (hunk 6 above)
Comment (nabijaczleweli) [inline server/src/gamestate.rs:19]: Clearly nonexistant yet

File: server/src/gamestate.rs (hunk 7 above)
Comment (EthanSteinberg) [inline server/src/gamestate.rs:53]: Yes.

**Comment by nabijaczleweli** [comment]
:+1:

**Comment by nabijaczleweli** [comment]
Reviewed 1 of 5 files at r4, 1 of 6 files at r6, 3 of 3 files at r10.
Review status: 5 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/20)_

**Pull request merged**